### PR TITLE
[refactor][ci] Build the docker image with docker-maven-plugin

### LINKED ISSUE #16652: Build the docker image without dockerfile-maven-plugin
As described in https://github.com/apache/pulsar/discussions/16642, dockerfile-maven-plugin doesn't support Apple M1 and the project itself is end-of-life.

For supporting developers using Apple M1 clip, we can migrate the building logic to other plugin. Here's some candidates:

* [docker-maven-plugin](https://github.com/fabric8io/docker-maven-plugin) is provided by fabric8io and actively maintained. However, the plugin logic is quite different from dockerfile-maven-plugin as the former may be over fruitful.
* maven-exec-plugin. dockerfile-maven-plugin is effectively a thin wrapper of `docker` commands. We may directly run those commands with maven-exec-plugin.
* A makefile. Apache SkyWalking uses a [Makefile](https://github.com/apache/skywalking/blob/master/Makefile) to hold logics about building docker image. It should be in the same direction with maven-exec-plugin but switches build system switch to Makefile.

cc @lhotari what do you think? At least I'd like to know the effectively executed `docker` commands so that I can run from command line interface and build those images manually.

### PR DESCRIPTION
**Pull request by tisonkun**
* This closes #16652.
* This closes #16099.
* This closes #16677.

- [x] `doc-not-needed`

### CODE REVIEW

**Comment by tisonkun** [comment]
Verify build successfully on macOS with M1 chip. Although a weird issue doesn't exist in the CI environment: https://github.com/apache/pulsar/pull/17129/files#r948602287. It should not be included in this patch, though.

**Comment by tisonkun** [comment]
cc @skyleaworlder you're welcome to review this patch. It's based on your work.

cc @merlimat @nicoloboschi @eolivelli @Shoothzj @michaeljmarshall @mattisonchao 

Build docker image succeed. I may retrigger other flaky tests when all steps finish.

**Comment by tisonkun** [comment]
All required tests pass now.

**Comment by skyleaworlder** [comment]
I think packaging part is pretty good now! Thank you very much for fixing my previous pending errors!

**Review by Shoothzj — APPROVED** [review]

File: docker/pulsar/pom.xml
Comment (nodece) [inline docker/pulsar/pom.xml:135]: By the way, I notice that there are redundant configurations here, and I think we can simplify them.

```xml
<executions>
      <execution>
        <id>default</id>
        <phase>package</phase>
        <goals>
          <goal>build</goal>
          <goal>push</goal>
        </goals>
        <configuration>
          <images>
            <image>
              <name>${docker.organization}/pulsar</name>
              <build>
                <contextDir>${project.basedir}</contextDir>
                <tags>
                  <tag>latest</tag>
                  <tag>${project.version}</tag>
                </tags>
              </build>
            </image>
          </images>
        </configuration>
      </execution>
</executions>
```

Then maybe we also need to update the https://github.com/apache/pulsar/blob/master/docker/publish.sh#L65

So like using `${docker.organization}/pulsar:latest` instead of `pulsar:latest`.

File: docker/pulsar-all/pom.xml
Comment (nodece) [inline docker/pulsar-all/pom.xml:143]: By the way, I notice that there are redundant configurations here, and I think we can simplify them.

```xml
<executions>
      <execution>
        <id>default</id>
        <phase>package</phase>
        <goals>
          <goal>build</goal>
          <goal>push</goal>
        </goals>
        <configuration>
          <images>
            <image>
              <name>${docker.organization}/pulsar</name>
              <build>
                <contextDir>${project.basedir}</contextDir>
                <tags>
                  <tag>latest</tag>
                  <tag>${project.version}</tag>
                </tags>
              </build>
            </image>
          </images>
        </configuration>
      </execution>
</executions>
```

Then maybe we also need to update the https://github.com/apache/pulsar/blob/master/docker/publish.sh#L66

So like using `${docker.organization}/pulsar-all:latest` instead of `pulsar-all:latest`.

**Review by nodece — CHANGES_REQUESTED** [review]

File: docker/pulsar/pom.xml
Comment (nodece) [reply docker/pulsar/pom.xml:135]: There needs the PMC confirmation.

File: tests/docker-images/latest-version-image/pom.xml
Comment (tisonkun) [reply tests/docker-images/latest-version-image/pom.xml:118]: Please read [the document of `docker-maven-plugin`](https://dmp.fabric8.io/#build-configuration). There's no `tag` goal. To perform tagging, this patch shows the correct way:

https://github.com/apache/pulsar/blob/5037bb1c59c6818772c845ea7909d32682dd6085/tests/docker-images/latest-version-image/pom.xml#L126-L129

File: tests/docker-images/java-test-image/pom.xml
Comment (tisonkun) [reply tests/docker-images/java-test-image/pom.xml:149]: See https://github.com/apache/pulsar/pull/17148#discussion_r949711928

File: docker/pulsar/pom.xml
Comment (tisonkun) [reply docker/pulsar/pom.xml:135]: This seems reasonable. I'm glad to integrate with this change if @sijie @merlimat @codelipenghui can confirm.

I found the original logic was introduced at https://github.com/apache/pulsar/pull/4705

File: docker/pulsar-all/pom.xml
Comment (tisonkun) [reply docker/pulsar-all/pom.xml:143]: See https://github.com/apache/pulsar/pull/17148#discussion_r949713557

**Comment by tisonkun** [comment]
Thanks for your detailed review @nodece! Comments addressed or replied.

**Comment by nodece** [comment]
> Thanks for your detailed review @nodece! Comments addressed or replied.

LGTM, I will approve this PR after `add-no-repo` is solved.

File: docker/pulsar/pom.xml
Comment (tisonkun) [reply docker/pulsar/pom.xml:111]: CI failed with:

```
Error:  Failed to execute goal io.fabric8:docker-maven-plugin:0.40.2:build (default) on project pulsar-docker-image: No image 'busybox' found and pull policy 'Never' is set. Please chose another pull policy or pull the image yourself) -> [Help 1]
```

After a few investigation, I notice that the original `dockerfile.build.pullNewerImage` means `Updates base images automatically`. So if we configure `imagePullPolicy` as `IfNotPresent` (default), we just retain the behaviour.

File: docker/pulsar/pom.xml
Comment (codelipenghui) [reply docker/pulsar/pom.xml:135]: We don't need to change the image publish process

Please see: https://github.com/apache/pulsar/blob/master/wiki/release/release-process.md#10-publish-docker-images

Comment (nodece) [reply docker/pulsar/pom.xml:135]: @codelipenghui I checked this process. Using the above config, we need to change the docker tag in `publish.sh`.

Comment (tisonkun) [reply docker/pulsar/pom.xml:135]: @nodece I think the point here is that it's possible that the release manager doesn't have the permission to push images to apachepulasr org. Thus we leave the ability to push to another org and ask the maintainer to sync. In this case, we generate the image with no repository for adding org when `publish.sh`.

Based on this information, I tend to keep the logic as is and if you have further improvement idea, we can start a new thread to discuss it.

Comment (nodece) [reply docker/pulsar/pom.xml:135]: Ok, we can make a new PR to improve that.

**Review by nodece — APPROVED** [review]
LGTM

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**
